plugins/audio-s5gen2/fu-audio-s5gen2-ble-device.c

```diff
@@ -0,0 +1,615 @@
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_recv(self, buf, sizeof(buf), &read_len, error))
+		return FALSE;
+
+	validate = fu_struct_qc_gaia_v3_upgrade_disconnect_ack_parse(buf, read_len, 0, error);
+	if (validate == NULL)
+		return FALSE;
+
+	return fu_qc_s5gen2_ble_device_notify_release(self, error);
+}
+
+static gsize
+fu_qc_s5gen2_ble_device_data_size(FuQcS5gen2Impl *impl, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(impl);
+	/* TODO: atm for GAIA only */
+	return self->mtu - FU_STRUCT_QC_DATA_SIZE - FU_QC_S5GEN2_GAIA_V3_HDR_SZ - 3;
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_get_api(FuQcS5gen2BleDevice *self, GError **error)
+{
+	guint8 buf[FU_QC_S5GEN2_BLE_DEVICE_BUFFER_SZ] = {0};
+	gsize read_len;
+	g_autoptr(GByteArray) req = fu_struct_qc_gaia_v3_api_req_new();
+	g_autoptr(GByteArray) resp = NULL;
+
+	fu_struct_qc_gaia_v3_api_req_set_vendor_id(req, self->vid_v3);
+
+	if (!fu_qc_s5gen2_ble_device_send(self, req->data, req->len, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_recv(self, buf, sizeof(buf), &read_len, error))
+		return FALSE;
+
+	resp = fu_struct_qc_gaia_v3_api_parse(buf, read_len, 0, error);
+	if (resp == NULL)
+		return FALSE;
+
+	self->api_version.major = fu_struct_qc_gaia_v3_api_get_major(resp);
+	self->api_version.minor = fu_struct_qc_gaia_v3_api_get_minor(resp);
+
+	if (self->api_version.major < 3) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "GAIA protocol %u.%u is not supported",
+			    self->api_version.major,
+			    self->api_version.minor);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* read the list of supported features from device */
+static gboolean
+fu_qc_s5gen2_ble_device_get_features(FuQcS5gen2BleDevice *self, gboolean next, GError **error)
+{
+	guint8 buf[FU_QC_S5GEN2_BLE_DEVICE_BUFFER_SZ] = {0};
+	gsize read_len;
+	g_autoptr(GByteArray) req = fu_struct_qc_gaia_v3_supported_features_req_new();
+	g_autoptr(GByteArray) resp = NULL;
+
+	fu_struct_qc_gaia_v3_supported_features_req_set_vendor_id(req, self->vid_v3);
+
+	fu_struct_qc_gaia_v3_supported_features_req_set_command(
+	    req,
+	    (next == FALSE) ? FU_QC_GAIA_V3_CMD_GET_SUPPORTED_FEATURES_REQ
+			    : FU_QC_GAIA_V3_CMD_GET_SUPPORTED_FEATURES_NEXT_REQ);
+
+	if (!fu_qc_s5gen2_ble_device_send(self, req->data, req->len, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_recv(self, buf, sizeof(buf), &read_len, error))
+		return FALSE;
+
+	resp = fu_struct_qc_gaia_v3_supported_features_parse(buf, read_len, 0, error);
+	if (resp == NULL)
+		return FALSE;
+
+	/* must be odd: header 5B + feature pairs */
+	if (!(read_len & 0x01)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "got incorrect features list");
+		return FALSE;
+	}
+
+	/* parse feature:version pairs */
+	for (gsize i = FU_STRUCT_QC_GAIA_V3_SUPPORTED_FEATURES_SIZE; i < read_len; i += 2) {
```

Comment:
is FU_STRUCT_QC_GAIA_V3_SUPPORTED_FEATURES_SIZE < FU_QC_S5GEN2_BLE_DEVICE_BUFFER_SZ? maybe adding `&& i < sizeof(buf)` to the conditional would help me sleep at night.

Reply:
Impossible or "impossible". But ok